File: src/main/java/dev/toastcie/mathtoastools/vector/intvector/Vector2Int.java
package dev.toastcie.mathtoastools.vector.intvector;

public class Vector2Int {
    int x;
    int y;

    public Vector2Int(int x, int y) {
        this.x = x;
        this.y = y;
    }

    //calculations methods

    public Vector2Int add(Vector2Int other) {
        return new Vector2Int(this.x + other.x, this.y + other.y);
    }

    public Vector2Int subtract(Vector2Int other) {
        return new Vector2Int(this.x - other.x, this.y - other.y);
    }

    public Vector2Int prod(int scalar) {
        return new Vector2Int(this.x * scalar, this.y * scalar);
    }

    public Vector2Int cross(Vector2Int other) {
        return new Vector2Int(this.x * other.y, this.y * other.x);
    }

    public int dot(Vector2Int other) {
        return this.x * other.x + this.y * other.y;
    }
}

File: src/main/kotlin/dev/toastcie/mathtoastools/vector/intvector/Vector2Int.kt
package dev.toastcie.mathtoastools.vector.intvector

import kotlin.math.sqrt

/**
 *  Represents a 2D vector with integer values
 *
 * @property x the x value of the vector
 * @property y the y value of the vector
 */
data class Vector2Int(var x: Int, var y: Int) {

    //calculations methods
    /**
     * Adds two vectors together
     *
     * @param other the vector to add
     * @return the sum of the two vectors
     */
    operator fun plus(other: Vector2Int): Vector2Int = Vector2Int(x + other.x, y + other.y)
    /**
     * Adds a value to each component of the vector
     *
     * @param other the value to add
     * @return the sum of the vector and the value
     */
    operator fun plus(other: Int): Vector2Int = Vector2Int(x + other, y + other)

    /**
     * Subtracts two vectors
     *
     * @param other the vector to subtract
     * @return the difference of the two vectors
     */
    operator fun minus(other: Vector2Int): Vector2Int = Vector2Int(x - other.x, y - other.y)
    /**
     * Subtracts a value to each component of the vector
     *
     * @param other the value to subtract
     * @return the difference of the vector and the value
     */
    operator fun minus(other: Int): Vector2Int = Vector2Int(x - other, y - other)

    /**
     * Multiplies two vectors
     *
     * @param other the vector to multiply
     * @return the product of the two vectors
     */
    operator fun times(other: Vector2Int): Vector2Int = Vector2Int(x * other.y, y * other.x)
    /**
     * Multiplies a value to each component of the vector
     *
     * @param other the value to multiply
     * @return the product of the vector and the value
     */
    operator fun times(other: Int): Vector2Int = Vector2Int(x * other, y * other)

    /**
     * Divides two vectors
     *
     * @param other the vector to divide
     * @return the quotient of the two vectors
     */
    infix fun dot(other: Vector2Int): Int = x * other.x + y * other.y

    //data methods
}